GSoC 2024: Internationalization Implementation for Zimit Frontend

Onyx2406 opened the pull request:
Fixes #46 
See also https://github.com/openzim/overview/issues/28

## Overview
This pull request introduces comprehensive internationalization (i18n) support to the Zimit frontend.

## What Has Been Done
- **Vue I18n Integration**: Implemented Vue I18n for managing translations within the application, with support for lazy-loaded language files.
- **Language Selection**: Added a language selection dropdown in the navbar, enabling users to switch languages dynamically. This feature is backed by automatic detection of the user's browser language, setting the application's default language accordingly.
- **English and French Translations**: Currently, the application supports English and French, with key UI elements and prompts available in both languages.

## What Is Left
- **TranslateWiki Integration**: Formal integration with TranslateWiki is pending.
- **Automated Browser Language Detection Improvement**: While initial support for detecting the browser's language is implemented, further testing and refinements is needed across different browsers and user configurations.
- **Preparation for TranslateWiki**: Organizing and documenting the translation file structure to facilitate future integration with TranslateWiki.
- **Expand Language Support**: Adding more languages based on community demand and contributions.

## Screenshots
![image](https://github.com/openzim/zimit-frontend/assets/32770175/92cf3924-6a79-49bf-95da-1428f03e0157)

kelson42 commented:
@Onyx2406 Thank you for your PR, here a few questions:
* Does it fix https://github.com/openzim/zimit-frontend/issues/46
* Please remove any code related to https://github.com/openzim/zimit-frontend/issues/28 and make an other PR.
* I would recommend to read https://github.com/openzim/overview/wiki/openZIM-code-contribution-workflow (if not already done)
* I general, for complex issues (like this one) we recommend to put a comment first explaining how you plan to solve the issue (this avoid to create large gap between expectation and PR)
* I have assigned a reviewer who will have a look during next week

Onyx2406 commented:
> @Onyx2406 Thank you for your PR, here a few questions:
> 
> * Does it fix [Interface should be multilingual #46](https://github.com/openzim/zimit-frontend/issues/46)
> * Please remove any code related to [Blacklist requests that are duplicates of existing resources or bound to fail #28](https://github.com/openzim/zimit-frontend/issues/28) and make an other PR.
> * I would recommend to read https://github.com/openzim/overview/wiki/openZIM-code-contribution-workflow (if not already done)
> * I general, for complex issues (like this one) we recommend to put a comment first explaining how you plan to solve the issue (this avoid to create large gap between expectation and PR)
> * I have assigned a reviewer who will have a look during next week

It does fix #46, but this PR only adds translations to the strings within this repository using Vue I18n. It doesn't translate the form details from Zimfarm. This PR only contains code to solve #46 and not #28. (I wanted to refer to https://github.com/openzim/overview/issues/28).

kelson42 commented:
@nikerabbit @Onyx2406 Do you know if this translation file can be indgested by Translatewiki?

Onyx2406 commented:
@kelson42, they are using a similar structure at https://github.com/mathjax/MathJax-i18n, so it should work.

rgaudin reviewed (changes requested):
Thank you; it's promising.

- We need RTL support
- I see that `human_size_limit` and `human_time_limit` are not i18n
- Languages list if static at the moment: requiring manually modifying two files to add a new locale. With translatewiki adding more all the time, we'd want something automatic

FYI (but it's a different PR) I believe there might be some strings in the backend (error message?) that would also require integration to completely close this

benoit74 commented on ui/src/locales/en.json:
```diff
@@ -0,0 +1,55 @@
+{
+    "alertFeedback": {
```
What is it used for?

benoit74 commented on ui/src/locales/en.json:
```diff
@@ -0,0 +1,55 @@
+        "serverConverting": "One of our servers is currently converting that URL into a nice ZIM file. Depending on the number of pages and resources available, it can be a matter of minutes, hours or even days! Please be patient.",
+        "emailNotification": "You can close this window. You will get an email notification when the task is completed."
+    },
+    "switchButton": {
```
What is it used for?

Onyx2406 replied:
It's not used anywhere currently, I will remove unnecessary strings.

benoit74 commented on ui/src/locales/en.json:
```diff
@@ -0,0 +1,55 @@
+        "loading": "Loading…"
+    },
+    "navbar": {
+        "logoAlt": "Youzim.it logo"
```
This has not been used in code while it should

Onyx2406 replied:
I will push the commits with these changes,

benoit74 reviewed (changes requested):
Very good job, I'm happy to see we are already quite close to have completed this task.

I think you've forgot to i18n the footer as well.

I propose to split the integration with TranslateWiki to another PR, we would already benefit from merging this PR once concerns have been addressed (Right To Left support is very important, I don't know if it works as of today, please test and confirm/adapt). @rgaudin are you ok with it?

Regarding @rgaudin comment about the automatic detection of new languages added, I do not think it is mandatory, especially it should not break lazy loading of translations. I do not mind to have to edit one more file everytime a new language is added (I imagine TranslateWiki folks might even be able to do it themselves when they add new languages).

Would you mind to create the "secret" `qqq` language where TranslateWiki expects to have translation instructions in English?

> TranslateWiki Integration: Formal integration with TranslateWiki is pending.

I will take care of that once PR is merged

> Automated Browser Language Detection Improvement: While initial support for detecting the browser's language is implemented, further testing and refinements is needed across different browsers and user configurations.

Do you have something special in mind to be tested? Or is it just "normal" testing?

> Preparation for TranslateWiki: Organizing and documenting the translation file structure to facilitate future integration with TranslateWiki.

I think the only thing left is to add the `qqq` magic language with translation instructions. Do you have something else in mind?

> FYI (but it's a different PR) I believe there might be some strings in the backend (error message?) that would also require integration to completely close this

Definitely a second step, let's finish this first one for now.

Nikerabbit commented:
> @Nikerabbit @Onyx2406 Do you know if this translation file can be indgested by Translatewiki?

Yes, we do support nested JSON. Also FYI we have recently updated our guidance for new projects: https://translatewiki.net/wiki/Translating:New_project

Onyx2406 commented:
Regarding splitting the integration with TranslateWiki into another PR, I completely agree. I have added the footer i18n.

I'll ensure that RTL support is tested and functional before merging. However, as the application currently supports only two languages which don't require RTL support, We can add RTL support to a future PR when additional RTL languages are added?

I've added the qqq language file for translation instructions.

For the automated browser language detection improvement, I'll do further testing across different browsers, so it's a normal testing and nothing special.

benoit74 commented:
Thank you ! Regarding RTL support, I do not mind to split it into another PR if there is too much work. It is however a pre-requisite for TranslateWiki integration and mostly a requirement for us, we have to translate the website into Farsi / Persian and already have persons ready to do the translation, so it will be in use quite quickly after TranslateWiki integration. Doesn't need to be perfect, but at least it should look pretty (e.g. I don't think we necessarily have to transfer all left buttons to the right and vice-versa, but at least the text should be readable and preferably right aligned when it was left aligned).

Onyx2406 commented:
Alright. In that case, it's better to push commits for RTL Support in this PR itself.

benoit74 commented on ui/src/components/NavBar.vue:
```diff
@@ -7,6 +7,7 @@
     <select @change="changeLanguage($event)" class="language-selector">
       <option value="en">English</option>
       <option value="fr">Français</option>
+      <option value="pers">Persian</option>
```
Value should by ISO639 so `far`, and label should be in Farsi so `فارسی`

benoit74 replied:
Nota: this means you have to update the file name ^^

benoit74 commented:
Btw, do not forget to have a look at `human_size_limit` and `human_time_limit`, this must be internationalized as well (especially for RTL I think). 

Regarding the persian translation, is it a language you speak or did you just performed an automated translation? If it is an automated translation, we will definitely need someone else to review it, I can't read it at all.

Nikerabbit commented:
FYI translatewiki.net prefers two letter language codes if one exists. Else we need to add code mapping.

Onyx2406 commented:
> just performed an automated translation

I had just performed an automated translation so we need someone to review it.

Onyx2406 commented:
Do we need to change the logo and buttons in RTL like this:

![image](https://github.com/openzim/zimit-frontend/assets/32770175/d2cf3714-21ef-40c3-8396-dbf025f305ee)

benoit74 commented:
@Onyx2406 I think this is exactly what we where referring to when mentioning adding support for RTL languages. 

@Popolechien @kelson42 @rgaudin do you confirm we want to reverse the layout when selecting a RTL language for Zimit as proposed in https://github.com/openzim/zimit-frontend/pull/51#issuecomment-1992289575?

Note that other screen(s) will have to be adapted as well.

rgaudin commented:
> @Popolechien @kelson42 @rgaudin do you confirm we want to reverse the layout when selecting a RTL language for Zimit 

Yes

Onyx2406 commented:
> Note that other screen(s) will have to be adapted as well.

@benoit74 Any screen you are referring to other than the main page? RTL support (reverses the layout) now works well for Farsi. It would be great if you could test it.

benoit74 commented:
> @benoit74 Any screen you are referring to other than the main page? RTL support (reverses the layout) now works well for Farsi. It would be great if you could test it.

I was referring to the fact that 3 components (and their subcomponents) have to be adapted: Request, NotFound and  NewRequest.

Please tell me once code is ready for a new review (or even better, request it directly through the dedicated button if you can, button is supposed to be in the upper right "Reviewers" box, I never recall if external contributors have access to it or not).

benoit74 reviewed (changes requested):
Thank you ! Still more stuff to fix unfortunately.

I don't get why `units.sizeLimit` is duplicated at `request.humanSizeLimit` and `units.timeLimit` is duplicated as `request.humanTimeLimit`. I suggest to keep only one at this stage (and prefer the `units` prefix which clearly indicates this is used in multiple screens).

There are some alignments issues on RTL:

![image](https://github.com/openzim/zimit-frontend/assets/7102089/69d2613b-af7d-49b3-950b-729ad3e2de76)

![image](https://github.com/openzim/zimit-frontend/assets/7102089/da5bb06f-9223-4774-81bf-4ee7f8f494bf)

![image](https://github.com/openzim/zimit-frontend/assets/7102089/913fa5b3-e16d-41af-9bd9-bced4d3219ff)

![image](https://github.com/openzim/zimit-frontend/assets/7102089/d7f584d6-04c4-4969-8051-1752c2da9a23)

Etc ... I think you got the point. Or is it due to the fact that my browser is not configured for a RTL language? I would prefer if we do not depend on this kind of browser setting.

The footer is not in proper order in RTL: Mozilla should be before - to the right - of Mozilla Open-source support award

![image](https://github.com/openzim/zimit-frontend/assets/7102089/f18a9d5d-1d9b-4e9a-b72c-25a684dbbeba)

Links in the FAQ are gone.

Pluralization is not handle appropriately (it wasn't handle before as well, but better to fix it now that I saw it):

![image](https://github.com/openzim/zimit-frontend/assets/7102089/08b3467d-1a78-4e8e-8076-e246b8acbde9)

I also realize we forgot to translate the email (in templates folder). Maybe we can do it in a second PR since this is backend code, to be done in Python. This makes me wonder how TranslateWiki will handle this, we will have to check with them.

benoit74 commented on ui/src/locales/en.json:
```diff
@@ -1,66 +1,67 @@
         "description": "Sorry, this content was not found. Maybe you'd like to go back."
     },
-    "request": {
-        "pending": "Pending",
```
Why did you removed all these translations?

benoit74 commented on ui/src/locales/en.json:
```diff
@@ -1,66 +1,67 @@
         "emailPlaceholder": "Your e-mail to receive a download link. Address not kept",
         "submit": "Let's Zim it!",
-        "advancedOptions": "advanced options",
-        "true": "True",
```
Why did you removed these translations?

benoit74 commented on ui/src/locales/fa.json:
```diff
@@ -1,78 +1,78 @@
 {
-    "faq": {
+    "faq": 
+    {
```
Why all in a sudden all `{` are moved to the next line? This is not standard JSON formatting, please revert (you should have `"faq": {` as before, and so on

benoit74 reviewed (changes requested):
Thank you, we are getting closer but you broke few things in your last commits (see conversations)

Please do not spend too much time fine-tuning the exact words of French and Farsi translations, since you've made them with an automated translation tool we will not merge them anyway, they will have to be translated by native speakers on TranslateWiki.

Other topics to address:
- you updated all locales but not the `qqq.json`, while it is probably the most important one
- why did you sometimes place HTML code inside the text to translate (e.g. `app.thanksToMozilla`) and other times you split the text is many substrings (e.g `faq.gotErrorDesc1` and `faq.gotErrorDesc2`) ? I believe it would be way simpler in most situations (e.g. not when HTML code is too complex, but in every other situations) to always have HTML code inside the translation and one single string to translate, it will make the translator work way easier rather than being presented with only few words out of any context, or even a dot or a comma to not forget at the beginning of a translation
  - e.g you should have only one `footer` string (`app` is a bit vague) and its value should be `Powered by <a target="_blank" href="https://kiwix.org">Kiwix</a> and <a target="_blank" href="https://webrecorder.net">Webrecorder</a>, thanks to a <a target="_blank" href="https://www.mozilla.org/moss/">Mozilla Open-Source Support</a> Award`
- limits in Request.vue JS code are still not handled properly:
  - they are not translated like in the Faq.vue
  - pluralization is not handled
- "+" buttons of the FAQ are still not on the left when RTL language is used
  - please have a look, if it is too complex I do not mind to move this to an issue for later fix, it is not a big usability concern

benoit74 commented on ui/src/components/FaqEntry.vue:
```diff
@@ -1,41 +1,54 @@
     top: .5em;
+  }
+
+  /* LTR specific styles */
```
Looks like it is possible to simplify by specifying both `left` and `right` value in all cases (ltr and rtl). See https://developer.mozilla.org/en-US/docs/Web/CSS/position?v=example#types_of_positioning and confirm if it works as expected on most browsers.

> If both left and right are specified, left wins when [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction) is ltr (English, horizontal Japanese, etc.) and right wins when [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction) is rtl (Persian, Arabic, Hebrew, etc.).

Onyx2406 replied:
Simplified it, works as expected.

benoit74 commented on ui/src/components/NewRequest.vue:
```diff
@@ -1,13 +1,13 @@
 <template>
   <div class="container">
-    <h1>Want an offline version of a website? Just <strong>Zim it</strong>!</h1>
+    <h1>{{$t('newRequest.heading')}} <strong>{{$t('newRequest.boldZim')}}</strong>!</h1>
```
This should be on single "thing" to translate, not two

benoit74 reviewed (changes requested):
Thank you, we are getting closer to merging this.

Still some comments in conversation plus few below.

- I just realized that the zimit logo was previously centered on the page while it is not left or right ; you must put it back to center (sorry to not have realized this before, I did not got it right)
- FYI, we will not integrate with TranslateWiki right away, I will open an issue with details on what still has to be done (but does not have to be addressed in this issue)

benoit74 commented:
New issue about what is left (not part of this issue / PR): https://github.com/openzim/zimit-frontend/issues/53

benoit74 reviewed (changes requested):
I'm sorry, looks like we didn't understood us well regarding the centering of the logo. I wanted to center only the logo, but not move the language button.

Expected UI in English:

![image](https://github.com/openzim/zimit-frontend/assets/7102089/c40cdaaf-9ab3-4728-97b7-4724598d8d0a)

Expected UI in Farsi:

![image](https://github.com/openzim/zimit-frontend/assets/7102089/cec3be4d-f6a4-4dfa-adfd-167fbe43eac6)

Rest is OK now

Onyx2406 commented on ui/src/components/NavBar.vue:
```diff
@@ -1,8 +1,10 @@
+    <router-link :to="{ name: 'home' }" class="logo">
+      <img alt="$t('navbar.logoAlt')" src="../assets/ZIMIT_LOGO_RGB.svg">
+    </router-link>
+    <Loading class="loading"/>
```
Removed it.

Onyx2406 commented on ui/src/components/NavBar.vue:
```diff
@@ -1,8 +1,10 @@
-    <router-link :to="{ name: 'home' }"><img alt="$t('navbar.logoAlt')" src="../assets/ZIMIT_LOGO_RGB.svg"></router-link>
-    <Loading/>
-
+  <header class="header">
```
Removed it.

Onyx2406 commented:
> Wouldn't it be simpler to move the select box out of the header, keep the header as it was and put an absolute positioning to the select box?

This worked, thanks!

benoit74 commented on ui/src/components/NavBar.vue:
```diff
@@ -1,24 +1,73 @@
+    border: 1px solid #ced4da;
+    appearance: none; 
+    background-color: #fff;
+    background-image: url('data:image/svg+xml;charset=UTF8,<svg ...></svg>'); /* Add a custom dropdown arrow */
```
What is this? That's a strange SVG isn't it?

benoit74 reviewed (changes requested):
One more thing ... Are you constantly generating code with a code generator? This is pretty annoying to spend so much time in reviews detecting code which means just nothing, or changes which are meaningless (random spaces added everywhere, ...)

benoit74 commented:
Thank you, unfortunately you again modified the indentation of many files.

I'm a bit sorry about that, but I will take over the end of this PR from this point, because I need to activate prettier and eslint on this, and this is way beyond the scope of this PR. 

Please consider that work is done from this point, I will finish what needs to be done. I will convert this to draft so that it is not merged by mistake.

Thank you.

benoit74 commented:
Done on our own finally
